### ISSUE TITLE
Custom DisplayNameFormatterExtension ignored after updating to 5.6.1

### ISSUE THREAD

**Issue by Jadarma** [kotest/kotest#3511]
I noticed that after upgrading to `5.6.1`, my custom `DisplayNameFormatter` no longer works.
It is present in the context registry but the functions are never called and the default formatting rules are used instead.
Reverting back to `5.5.5` fixes this issue.

**Kotest Version:** 5.6.1
**Kotlin Version:** 1.8.10
**Platform:** JVM / JUnit5

Minimal sample code:
```kotlin
internal class DummyNameFormatter : DisplayNameFormatter, DisplayNameFormatterExtension, Extension {
    override fun format(testCase: TestCase): String = testCase.name.testName + " (Custom)"
    override fun format(kclass: KClass<*>): String = kclass.simpleName.orEmpty() + " (Custom)"
    override fun formatter(): DisplayNameFormatter = this
}

object TestConfig : AbstractProjectConfig() {
    override fun extensions() = listOf<Extension>(DummyNameFormatter())
}

class MyTest : FunSpec({
    context("foo") {
        test("bar") {}
        test("baz") {}
    }
})
```

**Comment by myuwono** [comment]
@sksamuel it seems like the registration of the `AbstractProjectConfig` does not apply until the execution stage. The change introduced in #3461 means that on test discovery stage, it always use the default formatter. Do you have an idea on how to fix this? should we rethink how custom displayname formatters are registered?

**Comment by Jadarma** [comment]
Bug is still relevant. Bad bot!

**Comment by sksamuel** [comment]
Fixed in #3644
